File: ruby/brute_force.rb
require 'csv'
require 'benchmark'

# Measure the time taken to execute the code
time = Benchmark.measure do
  # Initialize variables to track the least expensive city and products
  city_prices = Hash.new(0)
  product_prices = Hash.new(Float::INFINITY)

  # Open the file for reading
  CSV.foreach("../ten_million_data.csv") do |row|
    city, product, price = row
    price = price.to_f

    # Update city prices
    city_prices[city] += price

    # Update product prices
    if price < product_prices[product]
      product_prices[product] = price
    end
  end

  # Find the least expensive city
  cheapest_city = city_prices.min_by { |_, total_price| total_price }

  # Find the 5 least expensive products
  cheapest_products = product_prices.sort_by { |_, price| price }.first(5)

  # Print the results
  puts "Ville la moins chère: #{cheapest_city[0]} #{format('%.2f', cheapest_city[1])}"
  puts "5 produits les moins chers:"
  cheapest_products.each do |product, price|
    puts "#{product} #{format('%.2f', price)}"
  end
end

# Print the time taken
puts "Temps écoulé: #{time.real} secondes"

File: js/brute_force.js
const fs = require('fs');
const readline = require('readline');

async function measureTime(callback) {
  const start = process.hrtime.bigint();
  await callback();
  const end = process.hrtime.bigint();
  const duration = Number(end - start) / 1_000_000_000; 
  console.log(`Temps écoulé: ${duration.toFixed(2)} secondes`);
}

measureTime(async () => {
  const fileStream = fs.createReadStream('../ten_million_data.csv');

  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  });

  const cityPrices = {};
  const productPrices = {};

  for await (const line of rl) {
    const [city, product, price] = line.split(',');
    const parsedPrice = parseFloat(price);

    if (!cityPrices[city]) {
      cityPrices[city] = 0;
    }
    cityPrices[city] += parsedPrice;

    if (!productPrices[product] || parsedPrice < productPrices[product]) {
      productPrices[product] = parsedPrice;
    }
  }

  let cheapestCity = null;
  let cheapestCityPrice = Infinity;
  for (const city in cityPrices) {
    if (cityPrices[city] < cheapestCityPrice) {
      cheapestCity = city;
      cheapestCityPrice = cityPrices[city];
    }
  }

  const cheapestProducts = Object.entries(productPrices)
    .sort((a, b) => a[1] - b[1])
    .slice(0, 5);

  console.log(`Ville la moins chère: ${cheapestCity} ${cheapestCityPrice.toFixed(2)}`);
  console.log('5 produits les moins chers:');
  cheapestProducts.forEach(([product, price]) => {
    console.log(`${product} ${price.toFixed(2)}`);
  });
});
